### PR TITLE
create better namespace objects

### LINKED ISSUE #7162: Object.assign used when importing json in development mode.
**Do you want to request a *feature* or report a *bug*?**

Bug

**What is the current behavior?**

In 'development' mode, importing JSON files generate `Object.assign` in the resulting bundle.

An example of this looks like: `Object.assign({}, _test__WEBPACK_IMPORTED_MODULE_1__, {\"default\": _test__WEBPACK_IMPORTED_MODULE_1__})`

This means that in development mode the resulting bundle does not work with IE (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign#Browser_compatibility)

**If the current behavior is a bug, please provide the steps to reproduce.**

Here's a sample repo to reproduce the issue: https://github.com/chaoyzeng/webpack-4-object-assign

You can check for the use of `Object.assign` in the resulting dev build bundle. Note that production mode does not use `Object.assign`

**What is the expected behavior?**

Don't use `Object.assign`, not sure what is the reasoning behind using it in the first place.

**Please mention other relevant information such as the browser version, Node.js version, webpack version, and Operating System.**

Webpack version is 4.6.0

### PR DESCRIPTION
**Pull request by sokra**
**What kind of change does this PR introduce?**
bugfix

**Did you add tests for your changes?**
yes

**If relevant, link to documentation update:**
N/A

**Summary**
also avoid using Object.assign (fixes #7162)

**Does this PR introduce a breaking change?**
bugfix

**What need to be documented?**
ESM namespace objects have `Symbol.toStringTag` set to `"Module"` -> https://tc39.github.io/ecma262/#sec-@@tostringtag

### CODE REVIEW

File: lib/MainTemplate.js [hunk 1]
```diff
@@ -253,11 +248,40 @@ module.exports = class MainTemplate extends Tapable {
 			buf.push(`${this.requireFn}.r = function(exports) {`);
 			buf.push(
 				Template.indent([
+					"if(typeof Symbol !== 'undefined' && Symbol.toStringTag) {",
```
Comment (TheLarkInn) [inline lib/MainTemplate.js]: Doesn't symbol break backwards IE compat?

Comment (TheLarkInn) [reply lib/MainTemplate.js]: If so we'd need make a breaking change.

Comment (paulfalgout) [reply lib/MainTemplate.js]: FWIW polyfilling `object.assign` for IE was all I needed to get it to function.

Comment (sokra) [reply lib/MainTemplate.js]: This PR is not only about Object.assign. It's about creating more spec compilant namespace objects for ESM and non-ESM interop.

Comment (TheLarkInn) [reply lib/MainTemplate.js]: Sorry I missed the typeof when I initially saw this.

**Pull request merged**